kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -212,22 +217,21 @@ else if ((state == State.START
                     && msg instanceof RequestFrame) {
                 bufferMsgAndSelectServer(msg);
             }
+            else if (state == State.CONNECTING || state == State.CONNECTED) {
+                bufferedMsgs.add(msg);
+            }
             else {
                 throw illegalState("Unexpected channelRead() message of " + msg.getClass());
             }
         }
     }
 
     private void bufferMsgAndSelectServer(Object msg) {
-        if (bufferedMsg != null) {
-            // Single buffered message assertion failed
-            throw illegalState("Already have buffered msg");
-        }
         state = State.CONNECTING;
         // But for any other request we'll need a backend connection
         // (for which we need to ask the filter which cluster to connect to
         // and with what filters)
-        this.bufferedMsg = msg;
+        this.bufferedMsgs.add(msg);
```

Comment:
The fact we do this in two places from the same if condition hints, to me at least, that we should extract a method for managing the buffer.